src/v/cluster/controller.cc

```diff
@@ -469,39 +477,91 @@ ss::future<> controller::stop() {
     });
 }
 
+ss::future<> controller::create_cluster() {
+    bootstrap_cluster_cmd_data cmd_data;
+    cmd_data.uuid = cluster_uuid(
+      uuid_t(boost::uuids::random_generator_mt19937()()));
+    vlog(clusterlog.info, "Creating cluster {}", cmd_data.uuid);
+
+    // cluster::replicate_and_wait() cannot be used here because it checks
+    // for serde_raft_0 feature to be enabled. Before a cluster is here,
+    // no feature is active, however the bootstrap_cluster_cmd is serde-only
+    return _stm
+      .invoke_on(
+        controller_stm_shard,
+        [this, cmd_data = std::move(cmd_data)](controller_stm& stm) mutable {
+            model::record_batch b = serde_serialize_cmd(
+              bootstrap_cluster_cmd{0, std::move(cmd_data)});
+            return stm.replicate_and_wait(
+              std::move(b),
+              model::timeout_clock::now() + 30s,
+              _as.local(),
+              std::nullopt);
+        })
+      .then([](std::error_code errc) {
+          vassert(errc == errc::success, "TODO: failed with {}", errc);
+      });
+}
+
 /**
  * This function provides for writing the controller log immediately
  * after it has been created, before anything else has been written
  * to it, and before we have started communicating with peers.
  */
-ss::future<> controller::cluster_creation_hook() {
-    if (!config::node().seed_servers().empty()) {
-        // We are not on the root node
-        co_return;
-    } else if (
-      _raft0->last_visible_index() > model::offset{}
-      || _raft0->config().brokers().size() > 1) {
-        // The controller log has already been written to
+ss::future<>
+controller::cluster_creation_hook(const bool local_node_is_seed_server) {
+    if (config::node().empty_seed_starts_cluster()) {
```

Comment:
Hmmm right we've been discussing that last week, and this is what is left from my original perspective. I think you're right about it but I just will think twice and will do the command driven bootstrap in both cases. In particular that would mean the original way of creating the initial user will go away, and even with `empty_seed_starts_cluster`, there will be `bootstrap_cluster_cmd` instead of `create_user_cmd`. That should work, though.